src/dbnode/storage/shard.go

```diff
@@ -1409,12 +1409,111 @@ func (s *dbShard) insertSeriesSync(
 	return newEntry, nil
 }
 
+func (s *dbShard) insertNewShardEntriesWithLock(entries []*Entry) {
+	if len(entries) == 0 {
+		return
+	}
+
+	// Fast Path: Check if the entire slice can be appended to the end of the list
+	if s.canAppendToEnd(entries[0]) {
+		s.appendEntriesToEnd(entries)
+		return
+	}
+
+	// If not, proceed with the standard insertion logic
+	elem := s.list.Back()
+	i := len(entries) - 1
+
+	for elem != nil && i >= 0 {
+		currListEntry := elem.Value.(*Entry)
+
+		insertIdx := s.findInsertionIndex(entries, currListEntry.Index, 0, i)
+
+		if insertIdx < len(entries) {
+			i = s.insertEntriesAfter(elem, entries, insertIdx, i)
+		}
+
+		elem = elem.Prev()
+	}
+
+	s.insertRemainingEntriesAtFront(entries, i)
+}
+
+// Helper function to check if we can append the entire slice to the end of the list
+func (s *dbShard) canAppendToEnd(firstEntry *Entry) bool {
```

Comment:
nit: all functions that assume you hold a Read lock should end with `...WithRLock` ; same thing for write lock with either `...WithLock` or `...WithWLock`

Reply:
Updated